File: Block Party-Objective_C/Block Party-Objective_C/Entities/Charm.m
//
//  Charm.m
//  Block Party-Objective_C
//
//  Created by Qi Feng Huang on 10/31/15.
//  Copyright © 2015 hackathon. All rights reserved.
//

#import "Charm.h"
#import "Constants.h"

@implementation Charm

//how fast charm moves across scene
static const NSTimeInterval MOVEMENT_SPEED = 2;

-(instancetype)init:(SKTexture *)texture worth:(int)worth parentSceneSize:(CGSize)size{
    //init with texture
    self = [super initWithTexture:texture color:[UIColor clearColor] size:texture.size];
    
    //set member variables
    self.worth = worth;
    self.zPosition = 75;
    
    //make charm smaller
    [self setScale:0.6];
    
    //create rectangular physics body
    SKPhysicsBody* physicsbody = [SKPhysicsBody bodyWithRectangleOfSize:self.size];
    
    //won't slow down when hit hero
    physicsbody.restitution = 0;
    physicsbody.friction = 0;
    
    //sets body type to CHARM
    physicsbody.categoryBitMask = CHARM;
    
    //charm won't collide with anything; HERO will do the collision checking
    physicsbody.collisionBitMask = NONE;
    physicsbody.contactTestBitMask = NONE;
    self.physicsBody = physicsbody;
    
    //set spawn location: top of scene
    [self setSpawnPosition: size];
    
    return self;
}

+(NSTimeInterval)getMovementSpeed{
    return MOVEMENT_SPEED;
}

-(SKAction*)getMovementAction{
    //moves to bottom side of scene and remove from parent
    SKAction* move = [SKAction moveToY:-self.size.height duration:MOVEMENT_SPEED];
    SKAction* remove = [SKAction removeFromParent];
    return [SKAction sequence:@[move, remove]];
}


-(void)setSpawnPosition:(CGSize)parentSceneSize{
    //set position at top of scene at random x position
    self.position = CGPointMake(randomPoint(0, parentSceneSize.width), self.size.height + parentSceneSize.height);
}
@end

File: Block Party-Swift/Block Party/Charm.swift
//
//  Charm.swift
//  Block Party
//
//  Created by Qi Feng Huang on 2/15/15.
//  Copyright (c) 2015 Qi Feng Huang. All rights reserved.
//

import SpriteKit

//collectable

class Charm: SKSpriteNode {
    
    let worth:Int //how much to add to score when collected
    
    static let moveSpeed:NSTimeInterval = 2
    
    init(spriteTexture:SKTexture, worth:Int){
        self.worth = worth
        super.init(texture: spriteTexture, color: UIColor.clearColor(), size: spriteTexture.size())
        zPosition = 75
        setScale(0.6)           //make a bit smaller
        
        //set up physics body
        let physicsbody = SKPhysicsBody(rectangleOfSize: CGSize(width: self.size.width, height: self.size.height))
        
        physicsbody.restitution = 0
        physicsbody.friction = 0
        physicsbody.categoryBitMask = PhysicsCategory.Charm
        physicsbody.collisionBitMask = PhysicsCategory.None
        physicsbody.contactTestBitMask = PhysicsCategory.None
        self.physicsBody = physicsbody

    }
    
    required init(coder aDecoder : NSCoder){
        fatalError("NSCoding not supported")
    }
    
    
    func getMoveAction()->SKAction{
        let move = SKAction.moveToY(-self.size.height, duration: Charm.moveSpeed)
        let remove = SKAction.removeFromParent()
        return SKAction.sequence([move,remove])
    }
    
    func getPosition(sceneSize: CGSize)->CGPoint{
        return CGPoint(x: randomPoint(min: 0, max: sceneSize.width), y: sceneSize.height + self.size.height)
    }
}